flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/matrix/SparseVector.java

```diff
@@ -0,0 +1,791 @@
+		int pos = Arrays.binarySearch(indices, i);
+		if (pos >= 0) {
+			return values[pos];
+		}
+		return 0.;
+	}
+
+	public void setSize(int n) {
+		this.n = n;
+	}
+
+	public int numberOfValues() {
+		return this.values.length;
+	}
+
+	@Override
+	public void set(int i, double val) {
+		int pos = Arrays.binarySearch(indices, i);
+		if (pos >= 0) {
+			this.values[pos] = val;
+		} else {
+			pos = -(pos + 1);
+			double[] newValues = new double[this.values.length + 1];
+			int[] newIndices = new int[this.values.length + 1];
+			int j = 0;
+			for (; j < pos; j++) {
+				newValues[j] = this.values[j];
+				newIndices[j] = this.indices[j];
+			}
+			newValues[j] = val;
+			newIndices[j] = i;
+			for (; j < this.values.length; j++) {
+				newValues[j + 1] = this.values[j];
+				newIndices[j + 1] = this.indices[j];
+			}
+			this.values = newValues;
+			this.indices = newIndices;
+		}
+	}
+
+	@Override
+	public void add(int i, double val) {
+		int pos = Arrays.binarySearch(indices, i);
+		if (pos >= 0) {
+			this.values[pos] += val;
+		} else {
+			pos = -(pos + 1);
+			double[] newValues = new double[this.values.length + 1];
+			int[] newIndices = new int[this.values.length + 1];
+			int j = 0;
+			for (; j < pos; j++) {
+				newValues[j] = this.values[j];
+				newIndices[j] = this.indices[j];
+			}
+			newValues[j] = val;
+			newIndices[j] = i;
+			for (; j < this.values.length; j++) {
+				newValues[j + 1] = this.values[j];
+				newIndices[j + 1] = this.indices[j];
+			}
+			this.values = newValues;
+			this.indices = newIndices;
+		}
+	}
+
+	@Override
+	public String toString() {
+		return "Sparse Vector{" +
+			"indices=" + Arrays.toString(indices) +
+			"values=" + Arrays.toString(values) +
+			"vectorSize=" + n +
+			'}';
+	}
+
+	@Override
+	public double normL2() {
+		double d = 0;
+		for (double t : values) {
+			d += t * t;
+		}
+		return Math.sqrt(d);
+	}
+
+	@Override
+	public double normL1() {
+		double d = 0;
+		for (double t : values) {
+			d += Math.abs(t);
+		}
+		return d;
+	}
+
+	@Override
+	public double normInf() {
+		double d = 0;
+		for (double t : values) {
+			d = Math.max(Math.abs(t), d);
+		}
+		return d;
+	}
+
+	@Override
+	public double normL2Square() {
+		double d = 0;
+		for (double t : values) {
+			d += t * t;
+		}
+		return d;
+	}
+
+	@Override
+	public SparseVector slice(int[] indices) {
+		SparseVector sliced = new SparseVector(indices.length);
+		int nnz = 0;
+		sliced.indices = new int[indices.length];
+		sliced.values = new double[indices.length];
+
+		for (int i = 0; i < indices.length; i++) {
+			int pos = Arrays.binarySearch(this.indices, indices[i]);
+			if (pos >= 0) {
+				sliced.indices[nnz] = i;
+				sliced.values[nnz] = this.values[pos];
+				nnz++;
+			}
+		}
+
+		if (nnz < sliced.indices.length) {
+			sliced.indices = Arrays.copyOf(sliced.indices, nnz);
+			sliced.values = Arrays.copyOf(sliced.values, nnz);
+		}
+
+		return sliced;
+	}
+
+	public SparseVector plus(SparseVector other) {
```

Comment:
Same comment from @becketqin in DenseVector. we should use a superClass `Vector` for the interface.

Reply:
Yes, we have made the change.